Lib/asyncio/base_events.py

```diff
@@ -601,29 +601,62 @@ def _check_running(self):
             raise RuntimeError(
                 'Cannot run the event loop while another loop is running')
 
-    def run_forever(self):
-        """Run until stop() is called."""
+    def run_forever_setup(self):
+        """Set up an event loop so that it is ready to start actively looping
+        to process events.
+
+        Returns the state that must be restored when the loop concludes. This state
+        should be passed in as arguments to ``run_forever_cleanup()``.
+
+        This method is only needed if you are writing your own event loop, with
```

Comment:
"is only needed if" -- odd phrasing. The method exists. Do you mean it only needs to be overridden? Maybe show a brief example of how it's supposed to be used in that case? (I intentionally didn't the docs before reading the code, and from only reading the docstring I'm a bit confused about what's hooking what. From reading more code I realize that setup/cleanup are what you intend to override.)

Reply:
No - it would be run_forever() that is being replaced. I've provided a sample implementation in the docs; I'll clarify the language here to (hopefully) make the intended usage more clear.